twilight-gateway/src/compression.rs

```diff
@@ -0,0 +1,181 @@
+            kind: CompressionErrorType::Decompressing,
+            source: Some(zstd_safe::get_error_name(code).into()),
+        }
+    }
+}
+
+impl fmt::Display for CompressionError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self.kind {
+            CompressionErrorType::Decompressing => f.write_str("message could not be decompressed"),
+            CompressionErrorType::NotUtf8 => f.write_str("decompressed message is not UTF-8"),
+        }
+    }
+}
+
+impl Error for CompressionError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        self.source
+            .as_ref()
+            .map(|source| &**source as &(dyn Error + 'static))
+    }
+}
+
+/// Type of [`CompressionError`] that occurred.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum CompressionErrorType {
+    /// Decompressing a frame failed.
+    Decompressing,
+    /// Decompressed message is not UTF-8.
+    NotUtf8,
+}
+
+#[cfg(feature = "zstd")]
+pub struct Decompressor {
+    /// Common decompressed message buffer.
+    buffer: Box<[u8]>,
+    /// Reusable zstd decompression context.
+    ctx: zstd_safe::DCtx<'static>,
+}
+
+/// Gateway event decompressor.
+#[cfg(feature = "zstd")]
+impl Decompressor {
+    /// [`Self::buffer`]'s size.
+    const BUFFER_SIZE: usize = 32 * 1024;
+
+    /// Create a new inflator for a shard.
+    pub fn new() -> Self {
+        Self {
+            buffer: vec![0; Decompressor::BUFFER_SIZE].into_boxed_slice(),
+            ctx: zstd_safe::DCtx::create(),
+        }
+    }
+
+    /// Decompress a message.
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`CompressionErrorType::Decompressing`] error type if the
+    /// message could not be decompressed.
+    ///
+    /// Returns a [`CompressionErrorType::NotUtf8`] error type if the
+    /// decompressed message is not UTF-8.
+    pub fn decompress(&mut self, message: &[u8]) -> Result<String, CompressionError> {
+        let mut input = zstd_safe::InBuffer::around(message);
+
+        // Decompressed message. `Vec::extend_from_slice` efficiently allocates
+        // only what's necessary.
+        let mut decompressed = Vec::new();
+
+        loop {
+            let mut output = zstd_safe::OutBuffer::around(self.buffer.as_mut());
+
+            self.ctx
+                .decompress_stream(&mut output, &mut input)
+                .map_err(CompressionError::from_code)?;
+
+            decompressed.extend_from_slice(output.as_slice());
+
+            // Break when message has been fully decompressed.
+            if input.pos == input.src.len() && output.pos() != output.capacity() {
+                break;
+            }
+        }
+
+        String::from_utf8(decompressed).map_err(CompressionError::from_utf8_error)
+    }
+
+    /// Reset the decompressor's internal state.
+    pub fn reset(&mut self) {
+        self.ctx
+            .reset(zstd_safe::ResetDirective::SessionOnly)
+            .expect("resetting session is infallible");
```

Comment:
I think this is good for now, but if we see memory grows over time or something we should switch it to just recreating the ctx.